<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CClust: Clustering Categorical Datasets • CClust</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CClust: Clustering Categorical Datasets">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CClust</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/CClust-vignette.html">CClust: Clustering Categorical Datasets</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>CClust: Clustering Categorical Datasets</h1>
                        <h4 class="author">R Package Version 0.0.0.9</h4>
            
            <h4 class="date">2019-06-15</h4>
      
      
      <div class="hidden name"><code>CClust-vignette.Rmd</code></div>

    </div>

    
    
<p><br></p>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><code>CClust</code> is an R package developed to help with clustering categorical datasets. The package can be installed from GitHub using devtools and then loaded in the usual way.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co"># Installs redres from GitHub</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"yudizhangzyd/CClust"</span>)</a></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># Loads the package</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(CClust)</a></code></pre></div>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>There are two main functions <code>kmodes</code> and <code>khaplotype</code> coded for clsutering categorical datasets, specifically the <code>khaplotype</code> function in this package can be used on clustering amplicon datasets with quality scores. In both <code>kmodes</code> and <code>khaplotype</code>, three k-means algorithms (Lloyd’s; MacQueen’s; Hartigan and Wong’s algorithm) were adapted to do clustering.</p>
</div>
<div id="package-structure" class="section level2">
<h2 class="hasAnchor">
<a href="#package-structure" class="anchor"></a>Package Structure</h2>
<p><code>CClust</code> is structured so the available functions fall under one of four categories.</p>
<ol style="list-style-type: decimal">
<li>
<a href="#read_data">Read data</a>: The first manner in which <code>CClust</code> can be used is to obtain amplicon datasets from a fastq file. This is done using the function <a href="#read_fastq"><strong><code>read_fastq</code></strong></a> to extract and return a list of data information: reads, quality scores and dimension of data.</li>
</ol>
<ol start="2" style="list-style-type: decimal">
<li>
<a href="#simulate">Simulate data</a>: The second way to utilize <code>CClust</code> is to is to Simulate clustered categorical datasets. This is done using the function <a href="#simulator"><strong><code>simulator</code></strong></a> to simulate clustered categorical datasets by using continous time Marcov chain.</li>
</ol>
<ol start="3" style="list-style-type: decimal">
<li>
<a href="#clustering">Clustering</a>: The third way to utilize <code>CClust</code> is to clustering categorical datasets using the following functions.</li>
</ol>
<p>-<a href="#khaplotype"><strong><code>khaplotype</code></strong></a>: Clustering the amplicon datasets with quality scores, only random initialization is avaiable in <code>khaplotype</code>.</p>
<p>-<a href="#kmodes"><strong><code>kmodes</code></strong></a>: Clustering categorical datasets without quality information of the data, different from function <code>khaplotype</code>, it has six different random initialization methods.</p>

<ol start="4" style="list-style-type: decimal">
<li>
<a href="#evaluation">Evaluation</a>: Lastly, <code>CClust</code> can be employed to evaluate the clustering result by the following functions.</li>
</ol>
<p>-<a href="#ARI"><strong><code>ARI</code></strong></a>: Computing the Adjusted Rand Index <span class="citation">(Hubert and Arabie 1985)</span> with the use of function <code>adjustedRandIndex</code> in package <code>mclust</code>.</p>
<p>-<a href="#plot_cluster"><strong><code>plot_cluster</code></strong></a>: Visulizing the clustering results with the use of function <code>dapc</code> in the package <code>adegenet</code>.</p>
<p>This vignette is set up to mimic the package structure. The sections after the introduction are dedicated to one of the three package uses. Additional details on the functions are included in these sections along with explanations of how to interpret the output and examples.</p>
</div>
<div id="algorithms" class="section level2">
<h2 class="hasAnchor">
<a href="#algorithms" class="anchor"></a>Algorithms</h2>
<p>Since both <code>kmodes</code> and <code>khaplotype</code> adapted three k-means algorithms (Lloyd’s; MacQueen’s; Hartigan and Wong’s algorithm), a brief review of these three algorithms are included here.</p>
<ol style="list-style-type: decimal">
<li>Lloyd’s algorithm <span class="citation">(Lloyd 1982)</span>
</li>
</ol>
<p>Lloyd’s algorithm prescribes alternating updates of the centers with updates of the cluster assignments. The algorithm is terminated if no obseravtion is transfered after a full circle of observations.</p>
<ol start="2" style="list-style-type: decimal">
<li>MacQueen’s algorithm <span class="citation">(MacQueen 1967)</span>
</li>
</ol>
<p>MacQueen suggested a slight modification on Lloyd’s algorithm: update the affected centers after every change to the cluster assignments. The algorithm is terminated if no obseravtion is transfered after a full circle of observations.</p>
<ol start="3" style="list-style-type: decimal">
<li>Hartigan and Wong’s algorithm <span class="citation">(Huang 1997)</span>
</li>
</ol>
<p>Hartigan’s algorithm improves on Lloyd’s and MacQueen’s algorithms by only assigning the observation to a new cluster if the objective function is guaranteed to improve. In particular, it takes into account any change in centers that will occur as a result of the move and makes sure that, even with this change, the value of the objective function will still be improved. Hartigan and Wong additionally define a live set to decide when to terminate the algorithm and to avoid unnecessary calculations. A cluster is no longer a live set only after a full cycle of the observations is made and none of them moved to or from that cluster.</p>
</div>
<div id="initialization-methods" class="section level2">
<h2 class="hasAnchor">
<a href="#initialization-methods" class="anchor"></a>Initialization methods</h2>
<p>For function <code>khaplotype</code>, only random initialization is implemented, users can set different seed.</p>
<p>For function <code>kmodes</code>, except for random initialization, there are 5 more initialization methods.</p>
<ol style="list-style-type: decimal">
<li>H97_RANDOM</li>
</ol>
<p>The first of two initialization schemes proposed by <span class="citation">Huang (1997)</span> uses the first K observations as the initial modes. This initialization scheme is randomized in the <code>R</code> package <code>klaR</code> by selecting K distinct modes at random from the n observations.</p>
<ol start="2" style="list-style-type: decimal">
<li>HD17</li>
</ol>
<p><span class="citation">Huang (1997)</span> mentioned above interpretted by Python author de Vos <span class="citation">(2017)</span>.</p>
<ol start="3" style="list-style-type: decimal">
<li>CLB09_RANDOM</li>
</ol>
<p>A deterministic initialization scheme peoposed by <span class="citation">Cao, Liang, and Bai (2009)</span>.</p>
<ol start="4" style="list-style-type: decimal">
<li>K-means++</li>
</ol>
<p>The k-means++ initialization scheme for k-means <span class="citation">Arthur and Vassilvitskii (2007)</span> adapted to k-modes.</p>
<ol start="5" style="list-style-type: decimal">
<li>Greedy K-means++</li>
</ol>
<p>A greedy version of the algorithm mentioned above <span class="citation">Arthur and Vassilvitskii (2007)</span>.</p>
</div>
<div id="example-data" class="section level2">
<h2 class="hasAnchor">
<a href="#example-data" class="anchor"></a>Example Data</h2>
<p><code>CCluster</code> contains three datasets, which will be used to demonstrate the functions in this vignette.</p>
<ol style="list-style-type: decimal">
<li><p>The <code>zoo.int.data</code> dataset is from the UCI Machine Learning Repository <span class="citation">(Lichman. 2013)</span>. It includes one numeric coordinate, number of legs, which is treated as categorical.</p></li>
<li><p>The <code>sim.fastq</code> dataset is sampled from a subset of an amplicon dataset mentioned in the paper DADA2 <span class="citation">(Callahan et al. 2016)</span>. Accordingly, the <code>assignment.txt</code> is the true assignment of that simulation data.</p></li>
<li><p>The <code>sim_small.fastq</code> dataset is a smaller simulated amplicon dataset from the same source mentioned above, it is for demonstrating the use of <code>plot_cluster</code>.</p></li>
</ol>
<p><br></p>
</div>
</div>
<div id="read_data" class="section level1">
<h1 class="hasAnchor">
<a href="#read_data" class="anchor"></a>Read Fastq File</h1>
<p>One of the ways to utilize <code>CClust</code> is to read the FASTQ file. This can be done using the function in the package named <code>read_fastq</code>.</p>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p><strong>Input</strong></p>
<ul>
<li>
<code>datafile</code>: Path to a fastq file.</li>
</ul>
<p><strong>Output</strong></p>
<ul>
<li>Return a list contains reads, quality socres (encoded using ASCII 33 to 126) and the dim of the data.</li>
</ul>
<p><strong>Functionality</strong></p>
<p>The code below demonstrates the use of <code>read_fastq</code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># Read the sim.fastq file</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">datFile &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"sim.fastq"</span>, <span class="dt">package =</span> <span class="st">"CClust"</span>)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">dat &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read_fastq.html">read_fastq</a></span>(<span class="dt">datafile =</span> datFile)</a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="co">#&gt; Minimum quality score: ( (40)</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="co">#&gt; Maximum quality score: G (71)</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="co">#&gt; Minimum read length: 251</span></a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="co">#&gt; Maximum read length: 251</span></a>
<a class="sourceLine" id="cb3-8" data-line-number="8"></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"><span class="co"># Prints the head of the data</span></a>
<a class="sourceLine" id="cb3-10" data-line-number="10">dat<span class="op">$</span>reads[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</a>
<a class="sourceLine" id="cb3-11" data-line-number="11"><span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]</span></a>
<a class="sourceLine" id="cb3-12" data-line-number="12"><span class="co">#&gt; [1,] "C"  "T"  "T"  "G"  "C"  "T"  "T"  "A"  "C"  "C"  </span></a>
<a class="sourceLine" id="cb3-13" data-line-number="13"><span class="co">#&gt; [2,] "C"  "T"  "T"  "G"  "C"  "T"  "T"  "A"  "C"  "C"  </span></a>
<a class="sourceLine" id="cb3-14" data-line-number="14"><span class="co">#&gt; [3,] "C"  "T"  "T"  "G"  "C"  "T"  "T"  "A"  "C"  "C"</span></a>
<a class="sourceLine" id="cb3-15" data-line-number="15">dat<span class="op">$</span>quality[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</a>
<a class="sourceLine" id="cb3-16" data-line-number="16"><span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]</span></a>
<a class="sourceLine" id="cb3-17" data-line-number="17"><span class="co">#&gt; [1,]   67   67   67   67   67   71   71   71   71    68</span></a>
<a class="sourceLine" id="cb3-18" data-line-number="18"><span class="co">#&gt; [2,]   67   67   67   67   67   71   71   71   71    71</span></a>
<a class="sourceLine" id="cb3-19" data-line-number="19"><span class="co">#&gt; [3,]   67   67   67   67   67   69   71   71   71    71</span></a>
<a class="sourceLine" id="cb3-20" data-line-number="20">dat<span class="op">$</span>dim</a>
<a class="sourceLine" id="cb3-21" data-line-number="21"><span class="co">#&gt; [1] 10000   251</span></a></code></pre></div>
<p><br></p>
</div>
</div>
<div id="simulator" class="section level1">
<h1 class="hasAnchor">
<a href="#simulator" class="anchor"></a>Simulate data</h1>
<p><code>CClust</code> can also be used to simulate clustered categorical datasets. This can be done using the function in the package named <code>simulator</code>.</p>
<div id="usage-1" class="section level2">
<h2 class="hasAnchor">
<a href="#usage-1" class="anchor"></a>Usage</h2>
<p><strong>Input</strong></p>
<ul>
<li>
<code>simK</code> Number of clusters.</li>
<li>
<code>n_coordinates</code> Number of coordinate of the simulated dataset.</li>
<li>
<code>n_observations</code> Number of observation of the simulated dataset.</li>
<li>
<code>sim_between_t</code> Between cluster variation.</li>
<li>
<code>sim_within_t</code> Within cluster variation.</li>
<li>
<code>use_dirichlet</code> Indicate if cimulate datasets with dirichlet prior. Defalut is FALSE.</li>
<li>
<code>sim_pi</code> Mixing proportions, a vector with the same length of specified number of clusters and the sum of the values in this vector has to be 1.</li>
</ul>
<p><strong>Output</strong></p>
<ul>
<li>Return a list of simulation dataset results, including the CTMC_probabilities, the true modes, the cluster_assignments, the cluster_sizes and simulated data.</li>
</ul>
<p><strong>Functionality</strong></p>
<p>The code below demonstrates the use of <code>simulator</code>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># Simulate data with dim 50 * 5, 4 different categories and there are 5 true clusters.</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">data &lt;-<span class="st"> </span><span class="kw"><a href="../reference/simulator.html">simulator</a></span>(<span class="dt">simK =</span> <span class="dv">5</span>, <span class="dt">n_coordinates =</span> <span class="dv">5</span>, <span class="dt">n_observations =</span> <span class="dv">50</span>, <span class="dt">n_categories =</span> <span class="dv">4</span>,</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="dt">sim_between_t =</span> <span class="dv">2</span>, <span class="dt">sim_within_t =</span> <span class="dv">1</span>, <span class="dt">use_dirichlet =</span> <span class="ot">TRUE</span>, <span class="dt">sim_pi =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="fl">0.1</span>, <span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.3</span>))</a>
<a class="sourceLine" id="cb4-4" data-line-number="4"></a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(data<span class="op">$</span>data)</a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5]</span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7"><span class="co">#&gt; [1,]    0    3    3    3    1</span></a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="co">#&gt; [2,]    0    3    2    3    1</span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="co">#&gt; [3,]    0    0    3    3    1</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10"><span class="co">#&gt; [4,]    1    2    3    3    1</span></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="co">#&gt; [5,]    0    0    3    0    1</span></a>
<a class="sourceLine" id="cb4-12" data-line-number="12"><span class="co">#&gt; [6,]    2    0    3    3    1</span></a></code></pre></div>
<p><br></p>
</div>
</div>
<div id="clustering" class="section level1">
<h1 class="hasAnchor">
<a href="#clustering" class="anchor"></a>Clustering categorical dataset</h1>
<p><code>CClust</code> has two clustering functions <code>kmodes</code> and <code>khaplotype</code>, <code>kmodes</code> is for clustering categorical datasets without quality information, <code>khaplotype</code> is for clustering datasets with quality information, now it only can be applied to amplicon datasets with quality scores.</p>
<div id="kmodes" class="section level2">
<h2 class="hasAnchor">
<a href="#kmodes" class="anchor"></a>kmodes</h2>
<p><code>kmodes</code> is intended to clustering the categorical datasets without quality information, it includes three unsupervised clustering algorithms and six initialization method.</p>
<p><strong>Inputs</strong></p>
<ul>
<li>
<code>K</code> Number of clusters. Default is 1.</li>
<li>
<code>datafile</code> Path to a data file.</li>
<li>
<code>n_init</code> Number of initializations.</li>
<li>
<code>algorithm</code> Algorithm to implement clustering. The following options are available: -<code>"KMODES_HUANG"</code>: MacQueen’s algorithm. (<em>default</em>) -<code>"KMODES_HARTIGAN_WONG"</code>: Hartigan and Wong algorithm. -<code>"KMODES_LLOYD"</code>: Lloyd’s algorithm.</li>
<li>
<code>init_method</code> Initialization methods. The following options are available: - <code>"KMODES_INIT_RANDOM_SEEDS"</code>: Random sampling. (<em>default</em>) - <code>"KMODES_INIT_H97_RANDOM"</code>: Huang1997, randomized version. - <code>"KMODES_INIT_HD17"</code>: Huang1997 interpretted by Python author de Vos. - <code>"KMODES_INIT_CLB09_RANDOM"</code>: Cao2009, randomized version. - <code>"KMODES_INIT_AV07"</code>: K-means++ adapted. - <code>"KMODES_INIT_AV07_GREEDY"</code>: K-means++ greedy adapted. - ``</li>
<li>
<code>seed</code> Random number seed. Default is 1.</li>
<li>
<code>shuffle</code> Incidate if shuffle the input order. Default is FALSE.</li>
</ul>
<p><strong>Output</strong></p>
<p>A list of clustering results. The following are returned detials: - <code>"best_cluster_size"</code>: Number of observations in each cluster of the best initialization. - <code>"best_criterion"</code>: Optimized criterion in each cluster of the best initialization. - <code>"best_cluster_id"</code>: Cluster assignment of each observation of the best initialization. - <code>"best_modes"</code>: Estimated modes for each cluster of the best initialization. - <code>"best_seed_index"</code>: Seed index of the best initialization. - <code>"total_best_criterion"</code>: Total optimized criterion of the best initialization. - <code>"clsuter_size"</code>: Number of clusters. - <code>"data_dim"</code>: Dimension of input data. - <code>"data"</code>: The input data.</p>
<p><strong>Functionality</strong></p>
<p>The code below shows how to use <code>kmodes</code> by changing types of algorithms and initialization methods. If run the algorithm more than one initialization, the <code>kmodes</code> function will return results of the initialization with the minimum <code>total_best_criterion</code>, which is the best solution among all of the initializations.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co"># Clustering with three initializations, default algorithm ("KMODES_HUANG") and shuffel the data with choosing K as 5</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">datFile &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"zoo.int.data"</span>, <span class="dt">package =</span> <span class="st">"CClust"</span>)</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">res_kmodes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/kmodes.html">kmodes</a></span>(<span class="dt">K =</span> <span class="dv">5</span>, <span class="dt">datafile =</span> datFile, <span class="dt">n_init =</span> <span class="dv">3</span>, <span class="dt">shuffle =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb5-4" data-line-number="4"></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="co"># Number of observations in each cluster</span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6">res_kmodes<span class="op">$</span>best_cluster_size</a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="co">#&gt; [1] 21 19 23 19 19</span></a></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co"># Clustering with Harigan and Wong and K-means++ greedy adapted initialization method.</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2">res_kmodes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/kmodes.html">kmodes</a></span>(<span class="dt">K =</span> <span class="dv">5</span>, <span class="dt">datafile =</span> datFile, <span class="dt">algorithm =</span> <span class="st">"KMODES_HARTIGAN_WONG"</span>, <span class="dt">init_method =</span> <span class="st">"KMODES_INIT_AV07_GREEDY"</span>)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"></a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="co"># Mode in each cluster</span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5">res_kmodes<span class="op">$</span>best_modes</a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13]</span></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="co">#&gt; [1,]    0    1    0    0    1    0    0    0    1     1     1     0     0</span></a>
<a class="sourceLine" id="cb6-8" data-line-number="8"><span class="co">#&gt; [2,]    6    0    0    1    0    0    0    1    0     0     1     0     0</span></a>
<a class="sourceLine" id="cb6-9" data-line-number="9"><span class="co">#&gt; [3,]    3    0    0    1    0    0    1    1    1     1     0     0     1</span></a>
<a class="sourceLine" id="cb6-10" data-line-number="10"><span class="co">#&gt; [4,]    0    0    0    0    1    0    1    1    1     1     1     0     1</span></a>
<a class="sourceLine" id="cb6-11" data-line-number="11"><span class="co">#&gt; [5,]    1    0    1    1    0    1    0    0    0     1     1     0     0</span></a>
<a class="sourceLine" id="cb6-12" data-line-number="12"><span class="co">#&gt;      [,14] [,15] [,16] [,17]</span></a>
<a class="sourceLine" id="cb6-13" data-line-number="13"><span class="co">#&gt; [1,]     2     1     0     1</span></a>
<a class="sourceLine" id="cb6-14" data-line-number="14"><span class="co">#&gt; [2,]     4     0     0     0</span></a>
<a class="sourceLine" id="cb6-15" data-line-number="15"><span class="co">#&gt; [3,]     0     1     0     0</span></a>
<a class="sourceLine" id="cb6-16" data-line-number="16"><span class="co">#&gt; [4,]     0     1     0     1</span></a>
<a class="sourceLine" id="cb6-17" data-line-number="17"><span class="co">#&gt; [5,]     1     1     0     0</span></a></code></pre></div>
</div>
<div id="khaplotype" class="section level2">
<h2 class="hasAnchor">
<a href="#khaplotype" class="anchor"></a>khaplotype</h2>
<p><code>khaplotype</code> is intended to clustering the categorical datasets with quality information, now it can be only used on amplicon dataset stored in a fastq file, it includes three unsupervised clustering algorithms and there is only one initialization method: random initialization. <code>khaplotype</code> can also be applied on the datasets that used for <code>kmodes</code> if set <code>run_with_quals</code> = FALSE, however it tends to be slower when implementing Hartigan and Wong’s algorithm.</p>
<p><strong>Inputs</strong></p>
<ul>
<li>
<code>K</code> Number of clusters. Default is 1.</li>
<li>
<code>datafile</code> Path to a data file.</li>
<li>
<code>n_init</code> Number of initializations.</li>
<li>
<code>algorithm</code> Algorithm to implement clustering. The following options are available: - <code>"FASTQ_LLOYDS_EFFICIENT"</code>: Efficient Lloyds algorithm - <code>"FASTQ_HW_EFFICIENT"</code>: Efficient Hartigan and Wong algorithm (<em>default</em>) - <code>"FASTQ_MACQUEEN"</code>: MacQueen’s algorithm - <code>"FASTQ_LLOYDS"</code>: Lloyds algorithm (Same as <code>"FASTQ_LLOYDS_EFFICIENT"</code> but slow, not suggested to use) - <code>"FASTQ_HW"</code>: Hartigan and Wong algorithm (Same as <code>"FASTQ_HW_EFFICIENT"</code> but slow, not suggested to use)</li>
<li>
<code>seed</code> Random number seed. Default is 1.</li>
<li>
<code>shuffle</code> Incidate if shuffle the input order. Default is FALSE.</li>
<li>
<code>run_with_quals</code> Indicate if conduct clustering on dataset with quality information. Default is TRUE.</li>
</ul>
<p><strong>Output</strong></p>
<p>A list of clustering results. The following are returned detials: - <code>"best_cluster_size"</code>: Number of observations in each cluster of the best initialization. - <code>"best_criterion"</code>: Optimized criterion in each cluster of the best initialization. - <code>"best_cluster_id"</code>: Cluster assignment of each observation of the best initialization. - <code>"best_modes"</code>: Estimated modes for each cluster of the best initialization. - <code>"total_best_criterion"</code>: Total optimized criterion of the best initialization. - <code>"clsuter_size"</code>: Number of clusters. - <code>"data_dim"</code>: Dimension of input data. - <code>"data"</code>: Reads of the input data.</p>
<p><strong>Functionality</strong></p>
<p>The code below shows how to use <code>khaplotype</code> by changing types of algorithms. If run the algorithm more than one initialization, the <code>khaplotype</code> function will return results of the initialization with the maximum <code>total_best_criterion</code>, which is the best solution among all of the initializations.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="co"># Clustering an amplicon dataset and run three initializations with default algorithm ("FASTQ_HW_EFFICIENT")</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2">datFile &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"sim.fastq"</span>, <span class="dt">package =</span> <span class="st">"CClust"</span>)</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">res_khap &lt;-<span class="st"> </span><span class="kw"><a href="../reference/khaplotype.html">khaplotype</a></span>(<span class="dt">K =</span> <span class="dv">5</span>, <span class="dt">datafile =</span> datFile, <span class="dt">n_init =</span> <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="co">#&gt; Minimum quality score: ( (40)</span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5"><span class="co">#&gt; Maximum quality score: G (71)</span></a>
<a class="sourceLine" id="cb7-6" data-line-number="6"><span class="co">#&gt; Minimum read length: 251</span></a>
<a class="sourceLine" id="cb7-7" data-line-number="7"><span class="co">#&gt; Maximum read length: 251</span></a>
<a class="sourceLine" id="cb7-8" data-line-number="8"><span class="co">#&gt; Time cost: 1.038276 secs</span></a>
<a class="sourceLine" id="cb7-9" data-line-number="9"><span class="co">#&gt; Log likelihood in 1th initialization: -108539.22 (5 iterations: 2979 4939)</span></a>
<a class="sourceLine" id="cb7-10" data-line-number="10"><span class="co">#&gt; Time cost: 0.491985 secs</span></a>
<a class="sourceLine" id="cb7-11" data-line-number="11"><span class="co">#&gt; Log likelihood in 2th initialization: -108624.68 (2 iterations: 5074 1)</span></a>
<a class="sourceLine" id="cb7-12" data-line-number="12"><span class="co">#&gt; Time cost: 0.628817 secs</span></a>
<a class="sourceLine" id="cb7-13" data-line-number="13"><span class="co">#&gt; Log likelihood in 3th initialization: -103563.00 (3 iterations: 2894 577)</span></a>
<a class="sourceLine" id="cb7-14" data-line-number="14"><span class="co">#&gt; Time cost is: 2.159193 secs</span></a>
<a class="sourceLine" id="cb7-15" data-line-number="15"><span class="co">#&gt; Best optimum is: -103562.999033</span></a>
<a class="sourceLine" id="cb7-16" data-line-number="16"></a>
<a class="sourceLine" id="cb7-17" data-line-number="17"><span class="co"># Number of observations in each cluster</span></a>
<a class="sourceLine" id="cb7-18" data-line-number="18">res_khap<span class="op">$</span>best_cluster_size</a>
<a class="sourceLine" id="cb7-19" data-line-number="19"><span class="co">#&gt; [1] 2894  577 1557  200 4772</span></a>
<a class="sourceLine" id="cb7-20" data-line-number="20"></a>
<a class="sourceLine" id="cb7-21" data-line-number="21"><span class="co"># The value of objective function of the best solutions</span></a>
<a class="sourceLine" id="cb7-22" data-line-number="22">res_khap<span class="op">$</span>total_best_criterion</a>
<a class="sourceLine" id="cb7-23" data-line-number="23"><span class="co">#&gt; [1] -103563</span></a></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="co"># Clustering an amplicon dataset and run three initializations with MacQueen's algorithm (shuffle the data)</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2">res_khap &lt;-<span class="st"> </span><span class="kw"><a href="../reference/khaplotype.html">khaplotype</a></span>(<span class="dt">K =</span> <span class="dv">5</span>, <span class="dt">datafile =</span> datFile, <span class="dt">n_init =</span> <span class="dv">3</span>, <span class="dt">algorithm =</span> <span class="st">"FASTQ_MACQUEEN"</span>, <span class="dt">shuffle =</span> <span class="ot">TRUE</span>, <span class="dt">seed =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb8-3" data-line-number="3"><span class="co">#&gt; Minimum quality score: ( (40)</span></a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="co">#&gt; Maximum quality score: G (71)</span></a>
<a class="sourceLine" id="cb8-5" data-line-number="5"><span class="co">#&gt; Minimum read length: 251</span></a>
<a class="sourceLine" id="cb8-6" data-line-number="6"><span class="co">#&gt; Maximum read length: 251</span></a>
<a class="sourceLine" id="cb8-7" data-line-number="7"><span class="co">#&gt; Time cost: 0.064216 secs</span></a>
<a class="sourceLine" id="cb8-8" data-line-number="8"><span class="co">#&gt; Log likelihood in 1th initialization: -197241.29 (2 iterations: 1552 4977)</span></a>
<a class="sourceLine" id="cb8-9" data-line-number="9"><span class="co">#&gt; Time cost: 0.060870 secs</span></a>
<a class="sourceLine" id="cb8-10" data-line-number="10"><span class="co">#&gt; Log likelihood in 2th initialization: -230540.18 (1 iterations: 4272 5607)</span></a>
<a class="sourceLine" id="cb8-11" data-line-number="11"><span class="co">#&gt; Time cost: 0.043078 secs</span></a>
<a class="sourceLine" id="cb8-12" data-line-number="12"><span class="co">#&gt; Log likelihood in 3th initialization: -200861.80 (1 iterations: 5070 17)</span></a>
<a class="sourceLine" id="cb8-13" data-line-number="13"><span class="co">#&gt; Time cost is: 0.168991 secs</span></a>
<a class="sourceLine" id="cb8-14" data-line-number="14"><span class="co">#&gt; Best optimum is: -197241.292293</span></a>
<a class="sourceLine" id="cb8-15" data-line-number="15"></a>
<a class="sourceLine" id="cb8-16" data-line-number="16"><span class="co"># Number of observations in each cluster</span></a>
<a class="sourceLine" id="cb8-17" data-line-number="17">res_khap<span class="op">$</span>best_cluster_size</a>
<a class="sourceLine" id="cb8-18" data-line-number="18"><span class="co">#&gt; [1] 1552 4977 2889  571   11</span></a></code></pre></div>
<p><br></p>
</div>
</div>
<div id="evaluation" class="section level1">
<h1 class="hasAnchor">
<a href="#evaluation" class="anchor"></a>Evaluation</h1>
<p>The final way to make use of <code>CClust</code> is to compute the adjusted rand index.</p>
<div id="ARI" class="section level2">
<h2 class="hasAnchor">
<a href="#ARI" class="anchor"></a>ARI</h2>
<p>The function <code>ARI</code> is to compute the adjusted rand index given the estimated assignments and the true assiganments.</p>
<p><strong>Inputs</strong></p>
<ul>
<li>
<code>est</code> A list of results returned from  or .</li>
<li>
<code>truth</code> A numeric or character vector of true assignemts.</li>
</ul>
<p><strong>Output</strong></p>
<ul>
<li>A numberic value between 0 and 1, which indicates agreement between two partitions.</li>
</ul>
<p><strong>Functionality</strong></p>
<p>The code below shows how to access the clustering results using function <code>ARI</code>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="co"># Estimate cluster assignments by function `khaplotype`.</span></a>
<a class="sourceLine" id="cb9-2" data-line-number="2">datFile &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"sim.fastq"</span>, <span class="dt">package =</span> <span class="st">"CClust"</span>)</a>
<a class="sourceLine" id="cb9-3" data-line-number="3">res_khap &lt;-<span class="st"> </span><span class="kw"><a href="../reference/khaplotype.html">khaplotype</a></span> (<span class="dt">K =</span> <span class="dv">5</span>, <span class="dt">datafile =</span> datFile, <span class="dt">n_init =</span> <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb9-4" data-line-number="4"><span class="co">#&gt; Minimum quality score: ( (40)</span></a>
<a class="sourceLine" id="cb9-5" data-line-number="5"><span class="co">#&gt; Maximum quality score: G (71)</span></a>
<a class="sourceLine" id="cb9-6" data-line-number="6"><span class="co">#&gt; Minimum read length: 251</span></a>
<a class="sourceLine" id="cb9-7" data-line-number="7"><span class="co">#&gt; Maximum read length: 251</span></a>
<a class="sourceLine" id="cb9-8" data-line-number="8"><span class="co">#&gt; Time cost: 1.021920 secs</span></a>
<a class="sourceLine" id="cb9-9" data-line-number="9"><span class="co">#&gt; Log likelihood in 1th initialization: -108539.22 (5 iterations: 2979 4939)</span></a>
<a class="sourceLine" id="cb9-10" data-line-number="10"><span class="co">#&gt; Time cost: 0.494279 secs</span></a>
<a class="sourceLine" id="cb9-11" data-line-number="11"><span class="co">#&gt; Log likelihood in 2th initialization: -108624.68 (2 iterations: 5074 1)</span></a>
<a class="sourceLine" id="cb9-12" data-line-number="12"><span class="co">#&gt; Time cost: 0.612562 secs</span></a>
<a class="sourceLine" id="cb9-13" data-line-number="13"><span class="co">#&gt; Log likelihood in 3th initialization: -103563.00 (3 iterations: 2894 577)</span></a>
<a class="sourceLine" id="cb9-14" data-line-number="14"><span class="co">#&gt; Time cost is: 2.128840 secs</span></a>
<a class="sourceLine" id="cb9-15" data-line-number="15"><span class="co">#&gt; Best optimum is: -103562.999033</span></a>
<a class="sourceLine" id="cb9-16" data-line-number="16">ass &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"assignment.txt"</span>, <span class="dt">package =</span> <span class="st">"CClust"</span>)</a>
<a class="sourceLine" id="cb9-17" data-line-number="17">true_assignments &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/numeric">as.numeric</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(ass, <span class="dt">header =</span> F, <span class="dt">sep =</span> <span class="st">""</span>))</a>
<a class="sourceLine" id="cb9-18" data-line-number="18"></a>
<a class="sourceLine" id="cb9-19" data-line-number="19"><span class="co"># Higher ARI means the similar between two partitions, so the clustering reasults is better</span></a>
<a class="sourceLine" id="cb9-20" data-line-number="20"><span class="kw"><a href="../reference/ARI.html">ARI</a></span>(res_khap, true_assignments)</a>
<a class="sourceLine" id="cb9-21" data-line-number="21"><span class="co">#&gt; [1] 0.9513036</span></a></code></pre></div>
</div>
<div id="Plotting" class="section level2">
<h2 class="hasAnchor">
<a href="#Plotting" class="anchor"></a>Plotting</h2>
<p>The function <code>plot_cluster</code> helps to visulize clusters after using the cluster algorithms implemented in the functions. This cluster plot make use of the function  from package  and function  from package . The plot is based on observation assignment and discriminant analysis of principal components, therefore, when running the code, it will asks you to choose the number of PCs and the number of discriminant functions to retain.</p>
<p><strong>Inputs</strong></p>
<ul>
<li>
<code>res</code> Results returned from <code>kmodes</code> or <code>khaplotype</code>.</li>
<li>
<code>xax</code> Integer specifying which principal components should be shown in x axes, default is 1.</li>
<li>
<code>yax</code> Integer specifying which principal components should be shown in y axes, default is 2.</li>
<li>
<code>isGene</code> Indicate if the clustering data is gene sequence.</li>
</ul>
<p><strong>Output</strong></p>
<ul>
<li>A plot reflecting clustering results.</li>
</ul>
<p><strong>Functionality</strong> The code below shows how to plot clusters after running the clustering algorithms using function <code>plot_cluster</code>.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="co"># read the data and use the function `kmodes` with default setting to do the clustering</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2">data &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"zoo.int.data"</span>, <span class="dt">package =</span> <span class="st">"CClust"</span>)</a>
<a class="sourceLine" id="cb10-3" data-line-number="3">res_kmodes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/kmodes.html">kmodes</a></span>(<span class="dt">K =</span> <span class="dv">5</span>, <span class="dt">datafile =</span> data, <span class="dt">algorithm =</span> <span class="st">"KMODES_HARTIGAN_WONG"</span>, <span class="dt">init_method =</span> <span class="st">"KMODES_INIT_AV07_GREEDY"</span>, <span class="dt">n_init =</span> <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb10-4" data-line-number="4"></a>
<a class="sourceLine" id="cb10-5" data-line-number="5"><span class="co"># Plot the clusters</span></a>
<a class="sourceLine" id="cb10-6" data-line-number="6"><span class="kw"><a href="../reference/plot_cluster.html">plot_cluster</a></span>(res_kmodes)</a></code></pre></div>
<p><img align="center" height="400" width="650" src="figs/kmodes_default.png"></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="co"># Plot the clusters by using other principle components</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="kw"><a href="../reference/plot_cluster.html">plot_cluster</a></span>(res_kmodes, <span class="dt">xax =</span> <span class="dv">2</span>, <span class="dt">yax =</span> <span class="dv">3</span>)</a></code></pre></div>
<p><img align="center" height="400" width="650" src="figs/kmodes_changed_pc.png"></p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="co"># read the data and use the function `khaplotype` with default setting to do the clustering</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2">data &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"sim_small.fastq"</span>, <span class="dt">package =</span> <span class="st">"CClust"</span>)</a>
<a class="sourceLine" id="cb12-3" data-line-number="3">res_khap &lt;-<span class="st"> </span><span class="kw"><a href="../reference/khaplotype.html">khaplotype</a></span>(<span class="dt">K =</span> <span class="dv">5</span>, <span class="dt">datafile =</span> data, <span class="dt">n_init =</span> <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb12-4" data-line-number="4"></a>
<a class="sourceLine" id="cb12-5" data-line-number="5"><span class="co"># Plot the clusters</span></a>
<a class="sourceLine" id="cb12-6" data-line-number="6"><span class="kw"><a href="../reference/plot_cluster.html">plot_cluster</a></span>(res_khap, <span class="dt">isGene =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p><img align="center" height="400" width="650" src="figs/khap.png"></p>
</div>
</div>
<div id="reference" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#reference" class="anchor"></a>Reference</h1>
<div id="refs" class="references">
<div id="ref-Arthur07">
<p>Arthur, D., and S. Vassilvitskii. 2007. “K-Means++: The Advantages of Careful Seeding. In Proceedings of the Eighteenth Annual Acm-Siam Symposium on Discrete Algorithms, New Orleans, Siam, Pp. 1027-1035.” <em>K-Means++: The Advantages of Careful Seeding</em>, January, 1027–35.</p>
</div>
<div id="ref-Benjamin2016">
<p>Callahan, B. J., P. J. McMurdie, Michael J Rosen, Andrew W Han, Amy Jo A Johnson, and Susan P Holmes. 2016. “DADA2: High-Resolution Sample Inference from Illumina Amplicon Data.” <em>Nature Methods</em> 13. Nature Publishing Group: 581–83.</p>
</div>
<div id="ref-Cao2009">
<p>Cao, F., J. Liang, and L. Bai. 2009. “A New Initialization Method for Categorical Data Clustering.” <em>Expert Syst. Appl.</em> 36: 10223–8.</p>
</div>
<div id="ref-Huang1997">
<p>Huang, Z. 1997. “A Fast Clustering Algorithm to Cluster Very Large Categorical Data Sets in Data Mining.” <em>Proceedings of the SIGMOD Workshop on Research Issues on Data Mining and Knowledge Discovery</em> 28: 1–8.</p>
</div>
<div id="ref-Hubert1985">
<p>Hubert, L., and P. Arabie. 1985. “Comparing Partitions.” <em>Journal of Classification</em>, 193–218.</p>
</div>
<div id="ref-Lichman13">
<p>Lichman., M. 2013. “UCI Machine Learning Repository, 2013.” <a href="http://archive.ics.uci.edu/ml">http://archive.ics.uci.edu/ml</a>.</p>
</div>
<div id="ref-Lloyd1982">
<p>Lloyd, S. 1982. “Least Squares Quantization in PCM.” <em>Information Theory, IEEE Transactions on</em> 28 (2): 129–37.</p>
</div>
<div id="ref-MacQueen1967">
<p>MacQueen, J. 1967. “Some Methods for Classification and Analysis of Multivariate Observations.” In <em>Proceedings of the Fifth Berkeley Symposium on Mathematical Statistics and Probability</em>, edited by L. M. Le Cam and J. Neyman, 1:281–97. Berkeley, CA: University of California Press.</p>
</div>
<div id="ref-Vos17">
<p>Vos, N. de. 2017. “Python K-Modes, Version 0.8.” <a href="https://github.com/nicodv/kmodes">https://github.com/nicodv/kmodes</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#introduction">Introduction</a><ul class="nav nav-pills nav-stacked">
<li><a href="#overview">Overview</a></li>
      <li><a href="#package-structure">Package Structure</a></li>
      <li><a href="#algorithms">Algorithms</a></li>
      <li><a href="#initialization-methods">Initialization methods</a></li>
      <li><a href="#example-data">Example Data</a></li>
      </ul>
</li>
      <li>
<a href="#read_data">Read Fastq File</a><ul class="nav nav-pills nav-stacked">
<li><a href="#usage">Usage</a></li>
      </ul>
</li>
      <li>
<a href="#simulator">Simulate data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#usage-1">Usage</a></li>
      </ul>
</li>
      <li>
<a href="#clustering">Clustering categorical dataset</a><ul class="nav nav-pills nav-stacked">
<li><a href="#kmodes">kmodes</a></li>
      <li><a href="#khaplotype">khaplotype</a></li>
      </ul>
</li>
      <li>
<a href="#evaluation">Evaluation</a><ul class="nav nav-pills nav-stacked">
<li><a href="#ARI">ARI</a></li>
      <li><a href="#Plotting">Plotting</a></li>
      </ul>
</li>
      <li><a href="#reference">Reference</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Yudi Zhang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
